cmd/stack_ls.go

```diff
@@ -78,69 +81,112 @@ func newStackLsCmd() *cobra.Command {
 				return stackSummaries[i].Name().String() < stackSummaries[j].Name().String()
 			})
 
-			_, showURLColumn := b.(httpstate.Backend)
-
-			// Devote 48 characters to the name width, unless there is a longer name.
-			maxName := 47
-			for _, summary := range stackSummaries {
-				name := summary.Name().String()
-				if len(name) > maxName {
-					maxName = len(name)
-				}
+			if jsonOut {
+				return formatJSON(b, current, stackSummaries)
 			}
-			maxName++ // Account for adding the '*' to the currently selected stack.
-
-			// Header string and formatting options to align columns.
-			formatDirective := "%-" + strconv.Itoa(maxName) + "s %-24s %-18s"
-			headers := []interface{}{"NAME", "LAST UPDATE", "RESOURCE COUNT"}
-			if showURLColumn {
-				formatDirective += " %s"
-				headers = append(headers, "URL")
-			}
-			formatDirective = formatDirective + "\n"
 
-			fmt.Printf(formatDirective, headers...)
-			for _, summary := range stackSummaries {
-				const none = "n/a"
+			return formatConsole(b, current, stackSummaries)
+		}),
+	}
+	cmd.PersistentFlags().BoolVarP(
+		&jsonOut, "json", "j", false, "Emit outputs as JSON")
+	cmd.PersistentFlags().BoolVarP(
+		&allStacks, "all", "a", false, "List all stacks instead of just stacks for the current project")
 
-				// Name column
-				name := summary.Name().String()
-				if name == current {
-					name += "*"
-				}
+	return cmd
+}
 
-				// Last update column
-				lastUpdate := none
-				if stackLastUpdate := summary.LastUpdate(); stackLastUpdate != nil {
-					lastUpdate = humanize.Time(*stackLastUpdate)
-				}
+func formatJSON(b backend.Backend, currentStack string, stackSummaries []backend.StackSummary) error {
+	type stackSummaryJSON struct {
+		Name          string `json:"name"`
+		Current       bool   `json:"current"`
+		LastUpdate    string `json:"lastUpdate,omitEmpty"`
+		ResourceCount *int   `json:"resourceCount,omitEmpty"`
+		URL           string `json:"url,omitEmpty"`
+	}
 
-				// ResourceCount column
-				resourceCount := none
-				if stackResourceCount := summary.ResourceCount(); stackResourceCount != nil {
-					resourceCount = strconv.Itoa(*stackResourceCount)
-				}
+	output := make([]stackSummaryJSON, len(stackSummaries))
+	for idx, summary := range stackSummaries {
+		summaryJSON := stackSummaryJSON{
+			Name:          summary.Name().String(),
+			ResourceCount: summary.ResourceCount(),
+			Current:       summary.Name().String() == currentStack,
+		}
+
+		if summary.LastUpdate() != nil {
+			summaryJSON.LastUpdate = summary.LastUpdate().Format(time.RFC3339)
```

Comment:
Do we want to print this as UTC time so that it is consistent across machines for the same stack data?

Reply:
Sounds good.  Done.